### DISCUSSION TITLE
RevealJS: Draw Shapes (Rectangle, Circle) in Slide at Absolute Positions

### DISCUSSION THREAD

**Discussion by RoyiAvital** [quarto-dev/quarto-cli#6845] [Q&A]
### Description

In [Slides](https://slides.com/) one can draw some shapes in the slide (Rectangles, Circles, etc...).

Is there a way, native or by a plug in, to draw shapes in Quarto as well?

**Comment by mcanouil** [comment]
You mean like these: <https://duckduckgo.com/?q=draw+shapes+in+revealjs>?

**Comment by RoyiAvital** [comment]
I am not sure what you mean.

But Slides has some shapes to add to the slides.
Since it is based on RevealJS I assume it is something can be done in Quarto as well.

To make it more concrete:

![image](https://github.com/quarto-dev/quarto-cli/assets/7347975/448493b9-74f2-4fe4-9bfd-27c5cb94d724)

Shapes in the left, then I could easily draw the red rectangle.

**Reply by mcanouil** [reply]
Did you look at the results which showed few plugins such as "canvas"?

**Reply by RoyiAvital** [reply]
I saw many results and probably due to location difference I saw different results than what you saw.

Any chance being more specific?

**Reply by mcanouil** [reply]
- https://evgladkiy.github.io/canvas/#/
- https://github.com/burnpiro/reveal-drawer

More generally, this repository is not the best place to search for Reveal.js plug-ins.
You better of looking on the Internet.

There are several Quarto extensions that embed such plugins, so you can easily make your own with the plugin of your choosing.

**Reply by RoyiAvital** [reply]
I was looking for a solution within Quarto.
I have no experience with RevealJS.

Also, the first project is using RevealJS but for explaining about HTML5 element.
The 2nd one is about drawing on top of the slides in viewing mode.

**Reply by mcanouil** [reply]
Well Quarto cannot and will not include every Reveal.js plug-ins.
The only way for you to get this kind of features is: to do it yourself or wait for someone in the community to do it.

Note that Quarto documentation is pretty much exhaustive on what it can "natively" do, see <https://quarto.org/docs/presentations/revealjs/> and <https://quarto.org/docs/reference/formats/presentations/revealjs.html>.

**Reply by mcanouil** [reply]
FYI, see the Reveal.js Quarto extension: <https://quarto.org/docs/extensions/listing-revealjs.html>

**Reply by RoyiAvital** [reply]
I am aware of those. Just thought there will be a quick / simple way to get away with this.
Let's see if there are other ideas.

**Reply by mcanouil** [reply]
> Just thought there will be a quick / simple way to get away with this.

If there is a Plugin that does what you want it's not embedded as a Quarto extension or completely unknown to me.
If it exists, you won't find it here but by looking on the internet.
Finally, if by "quick / simple way" you mean to wait for someone else to do it, I am afraid, it's not a very likely outcome.

> Let's see if there are other ideas.

I am not sure what other ideas you expect here ...

**Comment by jimjam-slam** [comment]
My guess is that the shapes buttons on slides.com probably draw SVG shapes. You could definitely include those SVG shapes in a Quarto revealjs presentation if you're willing to write the SVG yourself! Just put it inside an `{=html}` code block.

If you're looking for a visual interface to draw the shapes so that you don't have to write the SVG code yourself, you could use an SVG image editor like [Boxy](https://boxy-svg.com)! You just need it to give you the code to copy-and-paste into your Quarto doc—or, if it gives you an SVG file, open that file in Notepad, TextEdit or a code editor and copy-and-paste the code from there.

**Reply by RoyiAvital** [reply]
Great! I am using `draw.io` but its generated `SVG` looks bloated.
I will try Boxy!.

I wonder if there could be simple templates so it will be available easily for starters as well.

**Reply by mcanouil** [reply]
> I wonder if there could be simple templates so it will be available easily for starters as well.

That's called extensions, again make yours, see <https://quarto.org/docs/extensions/creating.html>.

For example:
- https://github.com/EmilHvitfeldt/quarto-revealjs-cinco-de-mayo
- https://github.com/EmilHvitfeldt/quarto-revealjs-earth
- https://quarto.org/docs/extensions/listing-revealjs.html
- https://github.com/mcanouil/awesome-quarto#presentations

**Reply by RoyiAvital** [reply]
@mcanouil , Please assume that had I known how to program them I would have.
I am looking for solutions which are for users of `Quarto`, not users and developers.

The answer by @jimjam-slam was a step forward.

**Reply by mcanouil** [reply]
The making of extensions and the tutorials to do so were improved a lot for users to be able to customise their experience with Quarto.
If with your knowledge the tutorial is out of reach, that's fine, but you should consider trying it and help us improve it.

You repeatedly ask for others to do something custom for you. Note that it is fine, but asking for custom things is not the same as making feature requests.
There are folks offering these kind of services, maybe if you don't want or can't do it yourself, then you could look for these type of services to get exactly what you want.


Quarto team (they will correct me if I am wrong) don't have the resources to make custom templates for users and don't intend to do so, thus the extensions mechanism.

**Reply by RoyiAvital** [reply]
I am a user.
I'd like to use Quarto instead of traditional tools for presentations (PowerPoint, Slides.com).

So I am expressing issues I'm having.
The idea is:

1. Hopefully others have encountered those and overcome, they can share.
2. There is an interest, so if there is a shortcoming, it might be addressed by capable ones.
3. It stays my own wish and others have little interest.

The path of programming it is known.
Yet it is not within my capabilities, unfortunately.

On many cases the community (You especially) has offered some reasonable solutions. 
So it all starts with sharing the issue.

The Quatro team always add features.
For instance, there is a template for `Callouts`. Why is that? Because there was a demand.
The Quatro team, probably has a roadmap of features.
It is also reasonable to assume their users have some effect on it.
This is why, I assume, they created the channels to communicate with them.

> Note that it is fine, but asking for custom things is not the same as making feature requests.

I start by asking if there are ways to do what I want.
I have read the documentation thoroughly, still it might be that I have missed something or there are some simple ways I'm not aware of. 
Hence I ask. If I get no answer, I will use the option for a _Feature Request_.

> You repeatedly ask for others to do something custom for you.

Can you give one example where I asked someone to do something?

I don't expect anything, I just communicate my needs as a user.
You repeatedly trying to educate me what to do and how.
If you can help with the question asked, please, I'd be grateful.
Otherwise, let it go. No need to teach me how to act.

**Comment by RoyiAvital** [comment]
I found some kind of a solution for rectangles and circles (And whatever is in between) based on a demo in the documentation: [RevealJS - Auto Animate - Element Matching](https://quarto.org/docs/presentations/revealjs/advanced.html#element-matching).

Basically, this becomes a constructor of a rectangle:

```qmd
::: {data-id="box1" style="background: #2780e3; width: 200px; height: 150px; margin: 10px;"}
Text inside
:::
```

I need to experiment to see if:
1. I can put it in arbitrary location using the `.absolute` class.
2. I can manipulate the text within it (Vertical alignment, Horizontal alignment, etc...).

**Reply by egavazzi** [reply]
If someone is looking for the same thing and find that discussion (like me): arbitrary position does work with @RoyiAvital solution. 

As an example:
```
::: {data-id="box1" style="background: #2780e3; width: 200px; height: 150px; margin: 10px; position: absolute; top: 400px; left: 200px"}
Text inside
:::
```

**Reply by 00113844** [reply]
any other ways to add other shapes?
